clients/src/main/java/org/apache/kafka/common/record/LazyDownConversionRecords.java

```diff
@@ -148,21 +149,24 @@ protected ConvertedRecords makeNext() {
                 return convertedBatch;
             }
 
-            if (!batchIterator.hasNext())
-                return allDone();
-
-            // Figure out batches we should down-convert based on the size constraints
-            List<RecordBatch> batches = new ArrayList<>();
-            boolean isFirstBatch = true;
-            long sizeSoFar = 0;
-            while (batchIterator.hasNext() &&
-                    (isFirstBatch || (batchIterator.peek().sizeInBytes() + sizeSoFar) <= maximumReadSize)) {
-                RecordBatch currentBatch = batchIterator.next();
-                batches.add(currentBatch);
-                sizeSoFar += currentBatch.sizeInBytes();
-                isFirstBatch = false;
+            while (batchIterator.hasNext()) {
+                List<RecordBatch> batches = new ArrayList<>();
+                boolean isFirstBatch = true;
+                long sizeSoFar = 0;
+
+                // Figure out batches we should down-convert based on the size constraints
+                while (batchIterator.hasNext() &&
+                        (isFirstBatch || (batchIterator.peek().sizeInBytes() + sizeSoFar) <= maximumReadSize)) {
+                    RecordBatch currentBatch = batchIterator.next();
+                    batches.add(currentBatch);
+                    sizeSoFar += currentBatch.sizeInBytes();
+                    isFirstBatch = false;
+                }
+                ConvertedRecords records = RecordsUtil.downConvert(batches, toMagic, firstOffset, time);
+                if (records.records().sizeInBytes() > 0)
```

Comment:
It's probably worth a comment here explaining which situations would result in 0 down-converted bytes.